### PR TITLE
Adaptions for Flutter Gallery demo

### PR DESCRIPTION
**Pull request by mosuem**
A series of changes for https://github.com/flutter/gallery/pull/1021.

- Make calls to loading a locale async
- Support `arb` files with different ordering and message keys than their reference file
- Generate constant hash
- Infer which `arb` file is the reference file
- Cleanups in generated code

After this PR, the plan is to publish experimental(!) versions of the packages to ease testing with other packages and applications.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by devoncarew** [comment]
nit and not necessarily related to this RP - we're now using `-wip` to signify there are changes to the packages since the last publish and we're not yet ready to re-publish (re: the `0.19.0-dev` tags above). The firehose tool special cases that extension and won't offer to publish a -wip version.

File: pkgs/messages/example_json/pubspec.yaml [hunk 1]
```diff
@@ -24,4 +24,4 @@ dev_dependencies:
 messages:
```
Comment (devoncarew) [inline pkgs/messages/example_json/pubspec.yaml:24]: For my own edification - we're using a `messages` key in the pubspec in order to configure package:messages_builder generation?

File: pkgs/messages_builder/lib/generation_options.dart [hunk 2]
```diff
@@ -18,24 +18,24 @@ class GenerationOptions {
...
-    required this.findByType,
+    required this.indexType,
   });
 
   static Future<GenerationOptions> fromPubspec(BuildStep buildStep) async {
     final pubspecId = await buildStep.findAssets(Glob('pubspec.yaml')).first;
     final pubspecData = await buildStep.readAsString(pubspecId);
     final pubspec = loadYaml(pubspecData) as YamlMap;
-    final messagesOptions = pubspec['messages'] as YamlMap;
+    final messagesOptions = pubspec['messages'] as YamlMap?;
```
Comment (devoncarew) [inline pkgs/messages_builder/lib/generation_options.dart:28]: Looking at the yaml files this is read from, the `messages:` key lacks a bit of context.

I do like that fact that we're not creating yet another data file in the user's directory (we have some to configure build, test, dartdoc, ...).

Should this key be `messages_builder:`? Should it be in a new 'options' section?

```
name: my_package
...

dependencies:
  path: any
  ...

options:
  messages_builder:
    use_method_names: true
```

Comment (devoncarew) [reply pkgs/messages_builder/lib/generation_options.dart:28]: I slightly prefer putting these new package options into a new category (`options:`), but realize there's a high degree of bike-shedding here.

File: pkgs/messages_builder/lib/message_parser/message_parser.dart [hunk 3]
```diff
@@ -12,11 +12,11 @@ import 'select_parser.dart';
 class MessageParser {
   static MessageWithMetadata parse(
     String debugString,
-    String fileContents, [
-    String? name,
+    String fileContents,
+    String name, [
     bool addId = false,
```
Comment (devoncarew) [inline pkgs/messages_builder/lib/message_parser/message_parser.dart:17]: This is existing code, but fyi, I almost always prefer named args over optional positional ones for the additional readability at the call site.

Comment (devoncarew) [reply pkgs/messages_builder/lib/message_parser/message_parser.dart:17]: (not sure whether the style guide speaks to this)

File: pkgs/messages_builder/pubspec.yaml [hunk 4]
```diff
@@ -10,13 +10,19 @@ environment:
...
   code_builder: ^4.3.0
+  collection: ^1.18.0
+  crypto: ^3.0.3
   dart_style: ^2.2.4
   glob: ^2.1.1
   intl: ^0.18.0
   messages:
-    path: ../messages
+    git:
+      url: git@github.com:dart-lang/i18n.git
```
Comment (devoncarew) [inline pkgs/messages_builder/pubspec.yaml:20]: Why the change to git references instead of local paths? We're using `dependency_overrides` (from a separate file) to satisfy local development. We'd want ~real pub.dev versions here for when we go to do a publish?

File: pkgs/messages_deserializer/pubspec.yaml [hunk 5]
```diff
@@ -9,11 +9,11 @@ environment:
 
 # Add regular dependencies here.
 dependencies:
-  code_builder: ^4.3.0
-  dart_style: ^2.2.4
   js: ^0.6.5
   messages:
-    path: ../messages
+    git:
```
Comment (devoncarew) [inline pkgs/messages_deserializer/pubspec.yaml:14]: Same question re: the git dep.

File: pkgs/messages_deserializer/pubspec.yaml [hunk 6]
```diff
@@ -9,11 +9,11 @@ environment:
...
+      url: git@github.com:dart-lang/i18n.git
+      path: pkgs/messages
 
 dev_dependencies:
   dart_flutter_team_lints: ^1.0.0
```
Comment (devoncarew) [inline pkgs/messages_deserializer/pubspec.yaml:19]: As a nit for a future PR - we're on on ^2.0.0 of dart_flutter_team_lints now.

File: pkgs/messages_shrinker/pubspec.yaml [hunk 7]
```diff
@@ -11,14 +11,23 @@ environment:
 dependencies:
   analyzer: ^6.2.0
   args: ^2.4.2
+  build: ^2.4.1
   messages:
-    path: ../messages
+    git:
+      url: git@github.com:dart-lang/i18n.git
```
Comment (devoncarew) [inline pkgs/messages_shrinker/pubspec.yaml:17]: So very many git dependencies.

**Review by devoncarew — COMMENTED** [review]
Generally looks good! A few questions, in-line in the PR.

> After this PR, the plan is to publish experimental(!) versions of the packages to ease testing with other packages and applications.

Lets publish to labs.dart.dev to start out -

File: pkgs/messages/example_json/pubspec.yaml (hunk 1 above)
Comment (mosuem) [reply pkgs/messages/example_json/pubspec.yaml:24]: Yes, I think it's better to keep the distinction between `package:messages` and `package:messages_builder` an implementation detail. This separation only exists to separate `dev_dependencies` from `dependencies`, which shouldn't concern the name.

File: pkgs/messages_builder/lib/generation_options.dart (hunk 2 above)
Comment (mosuem) [reply pkgs/messages_builder/lib/generation_options.dart:28]: I have no opinion on whether it should be in a new `options:` key, but it feels a bit redundant. What is the `pubspec` if not a collection of options? I took the inspiration from Flutter, where top-level keys are common I believe.

File: pkgs/messages_builder/pubspec.yaml (hunk 4 above)
Comment (mosuem) [reply pkgs/messages_builder/pubspec.yaml:20]: Yes, this was just to have the code actually run as long as the PR is not submitted.

File: pkgs/messages/example_json/pubspec.yaml (hunk 1 above)
Comment (mosuem) [reply pkgs/messages/example_json/pubspec.yaml:23]: On second thought, these options only concern the actual generation, so `messages_builder` might be nicer.

File: pkgs/messages_builder/pubspec.yaml (hunk 4 above)
Comment (devoncarew) [reply pkgs/messages_builder/pubspec.yaml]: Gotcha, understood.

File: pkgs/messages/example_json/pubspec.yaml (hunk 1 above)
Comment (devoncarew) [reply pkgs/messages/example_json/pubspec.yaml]: Thanks for the explanation re: the key; I'll comment on whether it should be nested or not on the other related comment.

File: pkgs/messages_builder/lib/generation_options.dart (hunk 2 above)
Comment (devoncarew) [reply pkgs/messages_builder/lib/generation_options.dart]: Flutter's are in a `flutter:` section; it feels large enough to get it's own section.

Mostly the top-level keys are for use by pub; it's less common for tools or packages to introduce new keys.

I think the important thing would be for somebody reading the file:

```
messages:
  generateMethods: true
  generateFindById: false
  generateFindBy: integer
```

to know that:
- these keys aren't mistaken for pub-related things
- are pretty clear what they're associated with (the package that reads them? the functionality used by the package under development - localization, message generation, ...)

`messages:` seems a little bare to be - slightly too little context. You might want play around with a few other keys / nested keys to see if something jumps out at you.

(I don't think it's important to change for this PR as this is all still WIP)

**Review by devoncarew — APPROVED** [review]
